Allow user to override sudo as root authorization method [minor revision requested]

eldering opened the pull request:
This is a followup on pull request #1558 and offers the user the option to preset the SUDO environment variable to override automatic detection of whether sudo, su or nothing (i.e. already root) should be used.

The reason is that not everyone (e.g. me) may have sudo rights installed, even if sudo is present.

erikrose commented on letsencrypt-auto:
```diff
@@ -61,10 +58,34 @@ if test "`id -u`" -ne "0" ; then
+}
+
+if /usr/bin/env | grep '^SUDO=' 1>/dev/null 2>&1; then
+  case "$SUDO" in
```
For consistency with the other env vars letsencrypt-auto checks, can we rename this `LE_AUTO_SUDO`? (Also, I share @pde's fear of an accidental collision with some other $SUDO hanging around.)

eldering replied:
I'm happy to change to another variable name. Should I create a new pull request?

erikrose commented on letsencrypt-auto:
```diff
@@ -35,14 +35,11 @@ done
-    echo \"sudo\" is not available, will use \"su\" for installation steps...
+# `su`. Auto-detection can be overrided by explicitly setting the
+# environment variable SUDO to 'sudo', 'sudo_su' or '' as used below.
+init_su_sudo() {
```
What is the purpose of this outer function? Is there a downside to unconditionally declaring `su_sudo()`?

eldering replied:
No, I don't see a downside. I think the main reason I wrote it like that was that it required less code line changes.

eldering commented:
I've addressed the comments of @erikrose in an updated pull request.

For some reason unclear to me the Travis build fails (at the end it seems) after a 10min timeout. Maybe this just needs a rebuild?

Could someone have a look at the changes and build failure and merge if it is acceptable now?

erikrose commented:
Restarted the bad Travis build. It looks like it was just a spurious timeout.

bmw commented:
Sorry this sat around so long @eldering. If you're still interested in getting this feature added to `{certbot,letsencrypt}-auto`, I'll review if you bring this PR up to date. To do this, you'll want to make changes to `letsencrypt-auto-source/letsencrypt-auto.template` and then run `letsencrypt-auto-source/build.py` to update `letsencrypt-auto-source/letsencrypt-auto`. All other `*-auto` files should remain unmodified as we only modify them during releases for legacy reasons.

I'm appending "[revision requested]" to the title of this PR which you should remove when you'd like a review.

eldering commented:
I refreshed my pull request: it is now applied to `letsencrypt-auto-source/lestencrypt-auto.template` at the top of master.
However, I need some help testing the behaviour properly: when I run
```
$ LE_AUTO_SUDO=su ./letsencrypt-auto
Using preset root authorization mechanism 'su'.
Upgrading certbot-auto 0.10.0.dev0 to 0.9.3...
Replacing certbot-auto...
Password: 
Password: 
Password: 
Requesting root privileges to run certbot...
  /home/jaap/.local/share/letsencrypt/bin/letsencrypt
[sudo] password for jaap: 
```
I assume that the self-updating feature overwrites my changes and it reverts to using `sudo`. I don't know where to look to fix this.

bmw commented:
Great. Thanks @eldering.

You're correct that `letsencrypt-auto` is updating itself and overwriting your changes. You can disable this self updating behavior with the `--no-self-upgrade` flag.

bmw commented on letsencrypt-auto-source/letsencrypt-auto.template:
```diff
@@ -84,39 +84,57 @@ fi
-  if command -v sudo 1>/dev/null 2>&1; then
-    SUDO=sudo
-    SUDO_ENV="CERTBOT_AUTO=$0"
+if /usr/bin/env | grep '^LE_AUTO_SUDO=' 1>/dev/null 2>&1; then
```
I think you can use `if [ -n "${LE_AUTO_SUDO+x}" ]; then` here.

bmw commented on letsencrypt-auto-source/letsencrypt-auto.template:
```diff
@@ -84,39 +84,57 @@ fi
+      SUDO=su_sudo
+      ;;
+    # Nothing to do for sudo and plain root methods
+    sudo) ;;
```
In the `sudo` case, you have to set `SUDO=sudo` and `SUDO_ENV="CERTBOT_AUTO=$0"` like we do below.

eldering commented:
I updated my pull request to incorporate your changes, thanks for these.

Unfortunately, I have not been able to fully test the 'su' option yet,
because when I run:
    LE_AUTO_SUDO=su ./letsencrypt-auto --no-self-upgrade
    Using preset root authorization mechanism 'su'.
    Bootstrapping dependencies for Debian-based OSes...
    Password:
    Hit http://security.debian.org jessie/updates InRelease
    Reading package lists... Done
    [...]
    Password:
    Reading package lists... Done
    Building dependency tree
    Reading state information... Done
    augeas-lenses is already the newest version.
    libaugeas0 is already the newest version.
    ca-certificates is already the newest version.
    gcc is already the newest version.
    libffi-dev is already the newest version.
    python is already the newest version.
    python-dev is already the newest version.
    python-virtualenv is already the newest version.
    virtualenv is already the newest version.
    libssl-dev is already the newest version.
    0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
    Using preset root authorization mechanism 'su'.
    Creating virtual environment...
    Installing Python packages...
    Had a problem while installing Python packages.
    Collecting argparse==1.4.0 (from -r
/tmp/tmp.bK9vnpS4Nz/letsencrypt-auto-requirements.txt (line 6))
      Downloading argparse-1.4.0-py2.py3-none-any.whl
    [...]
    Collecting certbot-nginx==0.9.3 (from -r
/tmp/tmp.bK9vnpS4Nz/letsencrypt-auto-requirements.txt (line 179))
      Downloading certbot_nginx-0.9.3-py2-none-any.whl (58kB)
    Requirement already satisfied (use --upgrade to upgrade):
setuptools>=11.3 in
/home/jaap/.local/share/letsencrypt/lib/python2.7/site-packages (from
cryptography==1.3.4->-r
/tmp/tmp.bK9vnpS4Nz/letsencrypt-auto-requirements.txt (line 36))
    Collecting python2-pythondialog>=3.2.2rc1 (from certbot==0.9.3->-r
/tmp/tmp.bK9vnpS4Nz/letsencrypt-auto-requirements.txt (line 173))
    In --require-hashes mode, all requirements must have their versions
pinned with ==. These do not:
        python2-pythondialog>=3.2.2rc1 from
https://pypi.python.org/packages/26/34/b4cf8018dbc51e8a0fae2c51a0dd2c1f34419caa9e3eb83646c73d1beb9d/python2-pythondialog-3.4.0.tar.gz#md5=554d611d435dcc072132586c1cb37ca5
(from certbot==0.9.3->-r
/tmp/tmp.bK9vnpS4Nz/letsencrypt-auto-requirements.txt (line 173))
    You are using pip version 8.0.3, however version 8.1.2 is available.
    You should consider upgrading via the 'pip install --upgrade pip'
command.

I don't know how to pin the python package versions.

Finally, for some reason I get a HTTP 500 error when trying to view
this specific pull request page now, but maybe that's just temporary.


On 31-10-16 17:03, Brad Warren wrote:
> *@bmw* requested changes on this pull request.
> 
> Other than these two minor comments, this LGTM. Thanks for the PR
> @eldering <https://github.com/eldering>!
> 
> ------------------------------------------------------------------------
> 
> In letsencrypt-auto-source/letsencrypt-auto.template
> <https://github.com/certbot/certbot/pull/1969#pullrequestreview-6501962>:
> 
>>  SUDO_ENV=""
>  export CERTBOT_AUTO="$0"
> -if test "`id -u`" -ne "0" ; then
> -  if command -v sudo 1>/dev/null 2>&1; then
> -    SUDO=sudo
> -    SUDO_ENV="CERTBOT_AUTO=$0"
> +if /usr/bin/env | grep '^LE_AUTO_SUDO=' 1>/dev/null 2>&1; then
> 
> I think you can use |if [ -n "${LE_AUTO_SUDO+x}" ]; then| here.
> 
> ------------------------------------------------------------------------
> 
> In letsencrypt-auto-source/letsencrypt-auto.template
> <https://github.com/certbot/certbot/pull/1969#pullrequestreview-6501962>:
> 
>>  SUDO_ENV=""
>  export CERTBOT_AUTO="$0"
> -if test "`id -u`" -ne "0" ; then
> -  if command -v sudo 1>/dev/null 2>&1; then
> -    SUDO=sudo
> -    SUDO_ENV="CERTBOT_AUTO=$0"
> +if /usr/bin/env | grep '^LE_AUTO_SUDO=' 1>/dev/null 2>&1; then
> +  case "$LE_AUTO_SUDO" in
> +    su_sudo|su)
> +      SUDO=su_sudo
> +      ;;
> +    # Nothing to do for sudo and plain root methods
> +    sudo) ;;
> 
> In the |sudo| case, you have to set |SUDO=sudo| and
> |SUDO_ENV="CERTBOT_AUTO=$0"| like we do below.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/certbot/certbot/pull/1969#pullrequestreview-6501962>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEeYNiR-KrnLS6oNqeMCUth-SQ6CP4j_ks5q5juZgaJpZM4G5Wps>.
>

bmw commented on letsencrypt-auto-source/letsencrypt-auto.template:
```diff
@@ -84,39 +84,59 @@ fi
+    sudo)
+      SUDO=sudo
+      SUDO_ENV="CERTBOT_AUTO=$0"
+	  ;;
```
It looks like spacing got messed up here. It'd be nice if this was inline with the other statements in the block.

bmw commented:
Travis failures aren't due to your PR. I'm looking into them now.

bmw commented:
I'm still at war with Travis. Closing and reopening to see if I can get tests to pass.

bmw commented:
LGTM. Thanks for the pr @eldering.

eldering commented:
It took a bit, but thanks for the effort to merge this!
